proto/provenance/exchange/v1/events.proto

```diff
@@ -0,0 +1,135 @@
+  uint64 order_id = 1;
+  // cancelled_by is the account that triggered the cancellation of the order.
+  string cancelled_by = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+
+// EventOrderFilled is an event emitted when an order has been filled in full.
+// This event is also used for orders that were previously partially filled, but have now been filled in full.
+message EventOrderFilled {
+  // order_id is the numerical identifier of the order filled.
+  uint64 order_id = 1;
+  // assets is the coins amount string of assets bought/sold for this order.
+  string assets = 2;
+  // price is the coins amount string of the price payed/received for this order.
+  string price = 3;
+  // fees is the coins amount string of settlement fees paid with this order.
+  string fees = 4;
+}
+
+// EventOrderPartiallyFilled is an event emitted when an order filled in part and still has more left to fill.
+message EventOrderPartiallyFilled {
+  // order_id is the numerical identifier of the order partially filled.
+  uint64 order_id = 1;
+  // assets is the coins amount string of assets that were filled and removed from the order.
+  string assets = 2;
+  // price is the coins amount string of the price payed/received for this order.
+  // For ask orders, this might be more than the amount that was removed from the order's price.
+  string price = 3;
+  // fees is the coins amount string of settlement fees paid with this partial order.
+  // For ask orders, this might be more than the amount that was removed from the order's settlement fees.
+  string fees = 4;
+}
+
+// EventMarketWithdraw is an event emitted when a withdrawal of a market's collected fees is made.
+message EventMarketWithdraw {
+  // market_id is the numerical identifier of the market.
+  uint32 market_id = 1;
+  // amount is the coins amount string of funds withdrawn from the market account.
+  string amount = 2;
+  // destination is the account that received the funds.
+  string destination = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // withdrawn_by is the account that requested the withdrawal.
+  string withdrawn_by = 4 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+
+// EventMarketDetailsUpdated is an event emitted when a market's details are updated.
+message EventMarketDetailsUpdated {
+  // market_id is the numerical identifier of the market.
+  uint32 market_id = 1;
+  // updated_by is the account that updated the details.
+  string updated_by = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+
+// EventMarketEnabled is an event emitted when a market is enabled.
+message EventMarketEnabled {
+  // market_id is the numerical identifier of the market.
+  uint32 market_id = 1;
+  // updated_by is the account that enabled the market.
+  string updated_by = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+}
+
+// EventMarketDisabled is an event emitted when a market is disabled.
+message EventMarketDisabled {
```

Comment:
We might be able to reduce the amount of events by changing the name to encompass both cases. The following is just an example:

```
message EventMarketStatusChange {
  // market_id is the numerical identifier of the market.
  uint32 market_id = 1;
  // updated_by is the account that enabled the market.
  string updated_by = 2 [(cosmos_proto.scalar) = "cosmos.AddressString"];
  // enabled is the status of the Market
  string enabled = 3 [(cosmos_proto.scalar) = "cosmos.AddressString"];
}
```

Reply:
I waffled on this quite a bit. I ultimately decided to go with the separate events because it puts the most important piece of information (whether it was enabled or disabled) in the event type which is more prominent than in an attribute. I also liked that, as separate event types, the events are a little smaller. These events probably won't be emitted often enough for the size differences to matter all that much, but it's still something to consider.

In general, you want events to be as small as possible. Usually, you don't want to include things (as attributes) that can be found with a query. You generally want the bare minimum amount of data that would allow you to a) recognize that something happened and b) be able to query for the details if you need them. With that in mind, I didn't really want a `bool enabled` in the event since you can query for the market to get its value. But I still really wanted the event to convey whether it was enabled or disabled, so I went with separate events whose names convey that information.

I also originally didn't have the `updated_by` field in there. Technically, it'd be possible to query for the info by getting the block txs, finding the right msg, and getting the field from it, but I decided that was a little too cumbersome. Plus, I imagine we (or someone) will need that `updated_by` information for an audit eventually.

But it's also nice to put things into events that aren't as easy to query for. Originally, the `EventOrderFilled` only had the order id. But for ask orders, the price received and fees paid depend on the bid(s) it was settled with and isn't something that can be actually queried. The actual price received might be split across multiple events generated by the bank module (since it might be coming from multiple accounts). So I felt it prudent to include those two fields in that event. I then added the `assets` field to it since the other info was there anyway, and now, something can use just those events to keep track of what things are being traded for.

Reply:
Events are in block results which is straight forward to process without needing to understand any of the proto msg formats or even run the "provenance app" (reading straight from tendermint/comet).

In my view you want events to be aggregated to some extent based on what you expect the listener to filter on ... a category of use.  It can be fine to add differential in json as empty/null attributes are omitted which doesn't cause excessive state bloat.  

Forcing related data to be queried breaks this model as then you need to query the "app" side of the blockchain process and you have associated issues with data retention (pruning intervals).  Relying on queries is a very bad place to be if you are event processing.

Reply:
Just to make sure we understand my concern... This was more about reducing the amount of code we have since we don't have any requirements by the consumer.